### PR TITLE
[0617] Migrate course updating subjects

### PR DESCRIPTION
**Pull request by defong**
### Context
Existing course subject updates

### Changes proposed in this pull request
Update course subjects

### Guidance to review

Navigate to 
1. the course listing page 
2. select an existing course
3. select basic details
4. on the subject row click on change

ie 
migrated implementation
https://teacher-training-api-pr-2570.london.cloudapps.digital/publish/organisations/2BD/2022/courses/39BP/details
vs original implementation
https://qa.publish-teacher-training-courses.service.gov.uk/organisations/2BD/2022/courses/39BP/details

The course
- name is regenerated
- ordering of the subject is enforced

For primary course flow
- select a `subject` as primary subject
- then save 

For secondary course flow
- select a `subject` as secondary subject
  - select second `subject` as secondary subject (optional)
- then save 

The modern language flow (intermediate steps will not save)
- select `modern language` as secondary subject
- select `named language`
- then save 

Not applicable for `further education` course

### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased master
- [x] Cleaned commit history
- [x] Tested by running locally

### CODE REVIEW

**Comment by defong** [comment]
[Migrate Courses - Course show page - Updating subjects](https://trello.com/c/S6m0PSHY/617-migrate-courses-course-show-page-updating-subjects)

**Review by mbocc — APPROVED** [review]

File: spec/services/courses/assign_subjects_service_spec.rb [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+  context "primary course" do
+    let(:subject_ids) { [primary_subject.id] }
+    let(:course) { Course.new(level: :primary) }
+    let(:primary_subject) { find_or_create(:primary_subject, :primary) }
+
+    it "sets the subjects" do
+      expect(subject.subjects.map(&:id)).to eq([primary_subject.id])
+    end
+
+    it "set the name" do
```
Comment (darokel) [inline spec/services/courses/assign_subjects_service_spec.rb:30]:
```suggestion
it "sets the name" do
```

File: spec/services/courses/assign_subjects_service_spec.rb [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  context "secondary course" do
+    let(:subject_ids) { [secondary_subject.id] }
+    let(:course) { Course.new(level: :secondary) }
+    let(:secondary_subject) { find_or_create(:secondary_subject, :biology) }
+
+    it "sets the subjects" do
+      expect(subject.subjects.map(&:id)).to eq([secondary_subject.id])
+    end
+
+    it "set the name" do
```
Comment (darokel) [inline spec/services/courses/assign_subjects_service_spec.rb:48]:
```suggestion
it "sets the name" do
```

File: spec/services/courses/assign_subjects_service_spec.rb [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+      it "sets the course subjects position" do
+        expect(subject.course_subjects.first.position).to eq(0)
+        expect(subject.course_subjects.first.subject.id).to eq(secondary_subject2.id)
+
+        expect(subject.course_subjects.second.position).to eq(1)
+        expect(subject.course_subjects.second.subject.id).to eq(secondary_subject.id)
+      end
+
+      it "set the name" do
```
Comment (darokel) [inline spec/services/courses/assign_subjects_service_spec.rb:72]:
```suggestion
it "sets the name" do
```

File: spec/services/courses/assign_subjects_service_spec.rb [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+  context "further_education course" do
+    let(:subject_ids) { nil }
+    let(:course) { Course.new(level: :further_education, provider: Provider.new) }
+    let(:further_education_subject) { find_or_create(:further_education_subject) }
+
+    it "sets the subjects" do
+      expect(subject.subjects.map(&:id)).to eq([further_education_subject.id])
+    end
+
+    it "set the name" do
```
Comment (darokel) [inline spec/services/courses/assign_subjects_service_spec.rb:87]:
```suggestion
it "sets the name" do
```

File: app/controllers/publish/courses/modern_languages_controller.rb [hunk 5]
```diff
@@ -4,7 +4,6 @@ class ModernLanguagesController < PublishController
       decorates_assigned :course
       before_action :build_course, only: %i[edit update]
       before_action :build_course_params, only: [:continue]
-      before_action :build_provider, only: [:new]
```
Comment (darokel) [inline app/controllers/publish/courses/modern_languages_controller.rb:7]: 🎉

File: app/controllers/publish/courses/modern_languages_controller.rb [hunk 6]
```diff
@@ -64,59 +64,49 @@ def current_step
...
 
-      def build_provider
-        @provider = RecruitmentCycle.find_by(year: params[:recruitment_cycle_year])
-                      .providers
-                      .find_by(provider_code: params[:provider_code])
+      def updated_subject_list
+        @updated_subject_list ||= selected_language_subjects_ids.concat(selected_non_language_subjects_ids)
+      end
+
+      def service
```
Comment (darokel) [inline app/controllers/publish/courses/modern_languages_controller.rb:71]: Could we rename this to reflect the class name

File: app/services/courses/assign_subjects_service.rb [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    def call
+      course.errors.add(:subjects, :duplicate) if request_has_duplicate_subject_ids?
+
+      update_subjects
+
+      course.name = course.generate_name
+      course
+    end
+
+    def subjects
```
Comment (darokel) [inline app/services/courses/assign_subjects_service.rb:21]: Do all the methods below `call` need to be private?

File: app/views/publish/courses/subjects/edit.html.erb [hunk 8]
```diff
@@ -0,0 +1,24 @@
+<% content_for :page_title, title_with_error_prefix("Change subjects – #{course.name_and_code}", course.errors.any?) %>
+
+<% content_for :before_content do %>
+  <%= govuk_back_link_to(details_publish_provider_recruitment_cycle_course_path(course.provider_code, course.recruitment_cycle_year, course.course_code)) %>
+<% end %>
+
```
Comment (darokel) [inline app/views/publish/courses/subjects/edit.html.erb:6]: Can we chop some of this spacing

File: spec/services/courses/assign_subjects_service_spec.rb (hunk 4 above)
Comment (defong) [reply spec/services/courses/assign_subjects_service_spec.rb]: done

File: app/services/courses/assign_subjects_service.rb (hunk 7 above)
Comment (defong) [reply app/services/courses/assign_subjects_service.rb:23]: done

File: app/controllers/publish/courses/modern_languages_controller.rb (hunk 6 above)
Comment (defong) [reply app/controllers/publish/courses/modern_languages_controller.rb]: done

**Review by darokel — APPROVED** [review]

**Pull request merged**